### ISSUE TITLE
Pyright ignores new instance attr type, if cast is done in __init__

### ISSUE THREAD

**Issue by ppiasek** [microsoft/pyright#10290]
**Describe the bug**
When using `cast` in child class `__init__`, pyright reports `reportAttributeAccessIssue` when accessing objects from the casted attribute.

It doesn't report issue, when the same cast is done in called method.

**Code**

```python
from typing import cast


class Param:
    pass


class ChildParam(Param):

    def child_method(self):
        pass


class Base:
    param: Param

    def __init__(self, param: Param):
        self.param = param


class Child(Base):
    def __init__(self, param: Param):
        super().__init__(param)
        self.param = cast(ChildParam, self.param)

    def method(self):
        self.param.child_method() # fail

    def method2(self):
        self.param = cast(ChildParam, self.param)
        self.param.child_method() # pass
```

**VS Code extension or command-line**
Pyright command line: `pyright 1.1.399`

**Comment by ppiasek** [comment]
Unfortunately, my real case here is a custom logger Handler using a custom Formatter. 
Custom Formatter has an additional method I want to call, because I want an additional formatting that returns a dictionary based on a log entry, which is then sent to the database. The logger itself is configured from json.

The best idea I found for pyright to not complain is cast. I don't think I can use Generics here

**Comment by erictraut** [comment]
Got it. Here's another potential workaround. It's not as clean as the generics solution and it's not entirely type safe, but I think it's safer and more maintainable than using `cast`. 

```python
class Child(Base):
    def __init__(self, param: ChildParam):
        super().__init__(param)
        self.param: ChildParam = self.param  # pyright: ignore[reportIncompatibleVariableOverride]
   ...
```
